dev-docs/RFCs/v7.x/layer-shader-hook-rfc.md

```diff
@@ -0,0 +1,107 @@
+
+### Example 2: Generic Brushing
+
+Brushing is based on mouse position and the position of the current object. To implement it in a way that is reusable across multiple layers, it requires the layers to expose shader hooks for object positions.
+
+### Example 3: Pattern Fill
+
+Instead of filling an area (triangle) with a solid color, fill it with a tiling pattern. The pattern would be supplied with a sprite image, with additional attributes controlling the size and the orientation. This is needed in rendering realistic maps e.g. a polygon or path that represents "crosswalk." This requires layers to expose shader hooks for object colors and texture coordinates.
+
+## Proposal
+
+### Naming Conventions
+
+Shader hook names follow the pattern `DECKGL_<action>_<target>`.
+
+- All hooks start with the `DECKGL_` prefix
+- Actions:
+  + `FILTER`: if the hook can modify the input value(s).
+  + `USE`: if the hook does not expect any output.
+- Targets:
+  + Colors are normalized to the WebGL color space.
+  + Coordinates are presumed to be in common space.
+  + Non-commonspace coordinates must contain modifiers in their names, e.g. `worldPosition`, `pixelSize`.
+
+### Standard Layer Hooks
+
+#### vs:DECKGL_FILTER_SIZE(inout vec3 size)
+
+Modify the commonspace dimensions associated with the current vertex, before it's projected.
+
+Semantic meanings:
+
+- ScatterplotLayer/IconLayer/PointCloudLayer/ColumnLayer: the offset from center
+- PathLayer/ArcLayer/LineLayer: the extrusion from the current vertex
+- SimpleMeshLayer: the local position
+- SolidPolygonLayer/BitmapLayer: (not applicable)
+
+
+#### vs:DECKGL_USE_POSITION(vec3 worldPosition)
+
+Observe the world space position associated with a point-like object.
+
+Semantic meanings:
+
+- SolidPolygonLayer (top): the vertex position (from `getPolygon`)
+- BitmapLayer: the vertex position (from `bounds`)
+- PathLayer/ArcLayer/LineLayer: (not applicable)
+- All other layers: the object center (from `getPosition`)
+
+
+#### vs:DECKGL_USE_POSITIONS(vec3 sourceWorldPosition, vec3 targetWorldPosition)
+
+Observe the world space source and target positions associated with an edge-like object.
+
+Semantic meanings:
+
+- ArcLayer/LineLayer: the source/target positions (from `getSourcePosition` and `getTargetPosition`)
+- PathLayer: the start/end of the current leg (from `getPath`)
+- SolidPolygonLayer (side): the two vertices that form the current edge (from `getPolygon`)
+- All other layers: (not applicable)
+
+
+#### vs:DECKGL_FILTER_COLOR(inout vec4 color, vec3 position, vec3 normal)
```

Comment:
is the position.w always `1` in all layers? so we skip it here?

Reply:
`w` is not always `1`. But your lighting module ignores it.